Fix downloading protobuf dependency

dcirne opened the pull request:
Protobuf has been failing due to its repository having dependencies. Downloading the .tar.gz form GitHub doesn't work because it does not bring its dependencies with it.

This PR introduces a function to clone a repository, initializa and update its submodules. The function is used to download protobuf as a repository dependency, rather than just a file.

The function is also implemented in a generic way, so if in the future other dependencies fall into the same situation, it can be reused.

With these changes, `tensorflow/workspace.bzl` goes back to using the commit sha in `PROTOBUF_URLS` and `PROTOBUF_STRIP_PREFIX`, rather than the tag of a release.

A command at the end of the `download_dependencies.sh` script was commented out. The comment says for it to be removed once protobug is fixed. Perhaps no longer necessary.

dcirne committed:
Fix downloading protobuf dependency

Protobuf has been failing due to its repository having dependencies. Downloading the .tar.gz form GitHub doesn't work because it does not bring its dependencies with it.

This PR introduces a function to clone a repository, initializa and update its submodules. The function is used to download protobuf as a repository dependency, rather than just a file.

The function is also implemented in a generic way, so if in the future other dependencies fall into the same situation, it can be reused.

With these changes, `tensorflow/workspace.bzl` goes back to using the commit sha in `PROTOBUF_URLS` and `PROTOBUF_STRIP_PREFIX`, rather than the tag of a release.

A command at the end of the `download_dependencies.sh` script was commented out. The comment says for it to be removed once protobug is fixed. Perhaps no longer necessary.

therc commented on tensorflow/workspace.bzl:
```diff
@@ -345,11 +345,11 @@ def tf_workspace(path_prefix = "", tf_repo_name = ""):
     )
 
     PROTOBUF_URLS = [
```
Perhaps add a comment that this is version 3.6.0. The SHA1 value is inscrutable.

dcirne committed:
Add comment to indicate the protobuf version

dcirne commented:
Conflict has been resolved. The PR is ready for review.

harshini-gadige commented:
@therc Can I proceed with next steps to merge this PR ? I don't see "approved" check and so want to get clarified on proceeding.

therc commented:
I'm not a tensorflow project member, I only added a drive-by comment. :)

harshini-gadige commented:
@dcirne  Waiting for the reviewer approval.
@petewarden  @wolffg  -  Any update ?

dcirne commented:
Resolved a conflict that appeared since our last contact.

dcirne commented:
@harshini-gadige with the news that 1.13 will be branching on Dec 13th, is there any chance this PR gets merged before that, in time to make the cut?

harshini-gadige commented:
> @dcirne Waiting for the reviewer approval.
> @petewarden @wolffg - Any update ?

@petewarden  @wolffg  -  Please review and approve this PR. If you are not the right person, please assign to the respective person.

penpornk reviewed (commented):
I'm so sorry for our delay! I've requested a review from @angersson who knows more about this than I do. In the meanwhile, your PR failed a couple of our tests. Could you please help take a look? Below are some quick links to the error log.
[Android Demo App](https://source.cloud.google.com/results/invocations/b812e3f2-60d7-4201-8d4a-ffef62914c7c/log)
[Ubuntu Makefile](https://source.cloud.google.com/results/invocations/7838a9ed-3790-48a4-8a81-577193f0d2c2/log)

Please ignore the `Windows Bazel GPU` failure. It's not related to this PR.

dcirne committed:
Update the Flatbuffer and Protobuf versions

dcirne commented:
Hi @penpornk , thank you for the links, they helped in debugging. It seems that the tests failing was not related to the content of the PR itself, but to the version of flatbuffer, which seems somehow dependent on protobuf version 3.6.1.2. Since This PR originally used protobuf version 3.6.1, the file couldn't be found. (See snippet from the tests below.)

```bash
downloading https://github.com/google/flatbuffers/archive/1f5eae5d6a135ff6811724f6c57f911d1f46bb15.tar.gz
650
Cloning into 'tensorflow/contrib/makefile/downloads/protobuf'...
651
error: pathspec 'v3.6.1.2.tar.gz' did not match any file(s) known to git.
652
================================================================================
653
//tensorflow/tools/ci_build/builds:gen_makefile_out                      FAILED in 0.1s
```

I updated the PR to use protobuf version 3.6.1.2. I hope this will fix the tests.

penpornk commented on tensorflow/contrib/makefile/download_dependencies.sh:
```diff
@@ -45,7 +47,7 @@ CUB_URL="$(grep -o 'https.*cub/archive.*zip' "${BZL_FILE_PATH}" | grep -v mirror
 # Required for TensorFlow Lite Flex runtime.
 FARMHASH_URL="https://mirror.bazel.build/github.com/google/farmhash/archive/816a4ae622e964763ca0862d9dbd19324a1eaf45.tar.gz"
-FLATBUFFERS_URL="https://github.com/google/flatbuffers/archive/1f5eae5d6a135ff6811724f6c57f911d1f46bb15.tar.gz"
+FLATBUFFERS_URL="https://github.com/google/flatbuffers/archive/v1.9.0.tar.gz"
```
Could you please change this back? I don't think it's related to the error.

penpornk replied:
The original URL also uses a newer version of flatbuffer (1.10.0). It was from [Aug 20](https://github.com/google/flatbuffers/commit/1f5eae5d6a135ff6811724f6c57f911d1f46bb15). The 1.9.0 release was in April.

dcirne committed:
Fix URLs for protobuf 3.6.1.2

dcirne commented:
@penpornk apologies for an earlier mistake. The commit I pushed about half an hour ago had an error in the URL to download version 3.6.1.2 of protobuf.

I have just corrected it and pushed a new commit. All the tests I could run on my side worked.

penpornk commented on tensorflow/contrib/makefile/download_dependencies.sh:
```diff
@@ -106,14 +131,16 @@ download_and_extract "${CUB_URL}" "${DOWNLOADS_DIR}/cub/external/cub_archive"
 # TODO(satok): Remove this once protobuf/autogen.sh is fixed.
-replace_by_sed 's#https://googlemock.googlecode.com/files/gmock-1.7.0.zip#http://download.tensorflow.org/deps/gmock-1.7.0.zip#' \
-  "${DOWNLOADS_DIR}/protobuf/autogen.sh"
+# replace_by_sed 's#https://googlemock.googlecode.com/files/gmock-1.7.0.zip#http://download.tensorflow.org/deps/gmock-1.7.0.zip#' \
```
Could you please explain why you commented this out? Is `protobuf/autogen.sh` fixed?

penpornk reviewed (changes requested):
@dcirne Thank you for your fast response! I think the problem is the setting of `PROTOBUF_SHA` itself. I tried the command on both Linux and Mac, both gave me `PROTOBUF_SHA=v3.6.1.2.tar.gz` instead of an actual commit hash.

As for why 3.6.1.2, it turns out your `tensorflow.bzl` was out of date. The latest `tensorflow.bzl` is already using protobuf 3.6.1.2. Please sync to the latest commit. :)

penpornk commented on tensorflow/contrib/makefile/download_dependencies.sh:
```diff
@@ -30,11 +30,13 @@ EIGEN_URL="$(grep -o 'http.*bitbucket.org/eigen/eigen/get/.*tar\.gz' "${BZL_FILE
+# These variables contain the GitHub repo and the sha, from `tensorflow/workspace.bzl`,
+# from which to clone it from and checkout to.
+readonly PROTOBUF_REPO="https://github.com/protocolbuffers/protobuf.git"
+readonly PROTOBUF_SHA="$(grep -o 'https://github.com/google/protobuf/archive/.*tar\.gz' "${BZL_FILE_PATH}" | head -n1 | awk '{print substr($0, index($0, "archive") + 8, 40) }')"
```
I got `PROTOBUF_SHA=v3.6.1.2.tar.gz` from this command instead of a commit hash.

penpornk reviewed (changes requested):
Forgot to highlight this line in my earlier review.

penpornk commented:
@dcirne I just saw your new comment after submitting my review. Sorry for not addressing your comment earlier! I'm rerunning the tests.

dcirne committed:
Updates from the PR review

`PROTOBUF_SHA` renamed to `PROTOBUF_TAG`. The variable is used to checkout to a specific tag after cloning the protobuf repository. The value is obtained by parsing the protobuf URL in the `workspace.bzl` file.

dcirne commented:
@penpornk thank you for the feedback. I have just pushed a new commit addressing all the points from your review.

penpornk reviewed (approved):
Thank you for the fixes! The changes look good to me. I'll go ahead and approve it. Hope the tests pass this time. :)

penpornk commented:
Only `Windows Bazel` and `Windows Bazel GPU` failed. These are existing failures. We are good to go. Thank you again for the PR!
